[Java][Flaky Test] CommandTests > scriptKill_unkillable (GlideClusterClient)

### Test Name

CommandTests > scriptKill unkillable (GlideClusterClient) > [1] RESP2

### Test Location

java/integTest - CommandTests

### Failure Permlink

https://github.com/valkey-io/valkey-glide/actions/runs/26556510434/job/78231053375#logs

### Frequency

Observed in scheduled CI run on 2026-05-28. Failed in 3 out of 3 Java jobs with EngineVersion 7.0 (Java 8, 11, 21 on x86_64-unknown-linux-gnu and container on amazon-linux).

### Steps to Reproduce

1. Run the FMT Java workflow on schedule
2. Observe failure in Java Tests with EngineVersion 7.0

### System Information

GitHub Actions (ubuntu, x86_64-unknown-linux-gnu)

### Language and Version

Java 8, 11, 21

### Engine Version

redis 7.0

### Logs

```
> Task :integTest:test
glide.cluster.CommandTests.scriptKill_unkillable(GlideClusterClient)[1]: FAILURE 8.105s

CommandTests > scriptKill unkillable (GlideClusterClient) > [1] RESP2 FAILED
    java.lang.AssertionError: Timed out waiting for 'unkillable' error after 8000ms
        at glide.cluster.CommandTests.lambda$pollForUnkillableInBackground$50(CommandTests.java:3757)
```

### Screenshots

_No response_

### Glide Version

main

### Expected Behavior

Test should pass reliably.

### Actual Behavior

scriptKill unkillable test fails consistently on redis 7.0 across multiple Java versions. Previously tracked in #5435 and #5476 (both closed).

### Possible Fixes

Investigate if the test has a race condition with script execution timing on redis 7.0.

--- a/java/integTest/src/test/java/glide/cluster/CommandTests.java
+++ b/java/integTest/src/test/java/glide/cluster/CommandTests.java
@@ -132,7 +132,7 @@ public class CommandTests {
 
     private static final String INITIAL_VALUE = "VALUE";
 
-    private static final int SCRIPT_POLL_TIMEOUT_MS = 8000;
+    private static final int SCRIPT_POLL_TIMEOUT_MS = 15000;
     private static final int SCRIPT_POLL_INTERVAL_MS = 500;
 
     private static final List<Arguments> clients = new ArrayList<>();
@@ -3943,13 +3943,13 @@ public class CommandTests {
 
         String key = UUID.randomUUID().toString();
         Route route = new SlotKeyRoute(key, PRIMARY);
-        String code = createLongRunningLuaScript(6, false);
+        String code = createLongRunningLuaScript(10, false);
 
         try (Script script = new Script(code, false);
                 GlideClusterClient testClient =
                         GlideClusterClient.createClient(
                                         commonClusterClientConfig()
-                                                .requestTimeout(10000)
+                                                .requestTimeout(15000)
                                                 .advancedConfiguration(
                                                         AdvancedGlideClusterClientConfiguration.builder()
                                                                 .connectionTimeout(10000)
@@ -4038,11 +4038,15 @@ public class CommandTests {
                                         result.complete(true);
                                         return;
                                     }
-                                    if (!msg.contains("no scripts in execution")) {
+                                    if (!msg.contains("no scripts in execution")
+                                            && !msg.contains("timed out")
+                                            && !msg.contains("timeout")) {
                                         // Unexpected error - fail fast
                                         result.completeExceptionally(e);
                                         return;
                                     }
+                                    // Expected: script hasn't started yet (NotBusy) or server is
+                                    // blocked by script (timeout) - keep polling
                                 }
                                 try {
                                     Thread.sleep(SCRIPT_POLL_INTERVAL_MS);

--- a/java/integTest/src/test/java/glide/cluster/CommandTests.java
+++ b/java/integTest/src/test/java/glide/cluster/CommandTests.java
@@ -24,6 +24,7 @@ import static glide.TestUtilities.getUnixSeconds;
 import static glide.TestUtilities.getValueFromInfo;
 import static glide.TestUtilities.isWindows;
 import static glide.TestUtilities.parseInfoResponseToMap;
+import static glide.TestUtilities.waitFor;
 import static glide.TestUtilities.waitForNotBusy;
 import static glide.TestUtilities.waitForSaveNotInProgress;
 import static glide.api.BaseClient.OK;
@@ -132,7 +133,7 @@ public class CommandTests {
 
     private static final String INITIAL_VALUE = "VALUE";
 
-    private static final int SCRIPT_POLL_TIMEOUT_MS = 15000;
+    private static final int SCRIPT_POLL_TIMEOUT_MS = 8000;
     private static final int SCRIPT_POLL_INTERVAL_MS = 500;
 
     private static final List<Arguments> clients = new ArrayList<>();
@@ -3943,13 +3944,25 @@ public class CommandTests {
 
         String key = UUID.randomUUID().toString();
         Route route = new SlotKeyRoute(key, PRIMARY);
-        String code = createLongRunningLuaScript(10, false);
+        String code = createLongRunningLuaScript(6, false);
+
+        // Lower lua-time-limit on the target primary so the server starts reporting BUSY (and
+        // accepting SCRIPT KILL) shortly after the script begins, instead of after the 5000ms
+        // default. Without this, SCRIPT KILL is queued behind the running script and blocks until
+        // the request timeout fires, which is the source of the flakiness.
+        String originalLuaTimeLimit =
+                clusterClient
+                        .configGet(new String[] {"lua-time-limit"}, route)
+                        .get()
+                        .getSingleValue()
+                        .get("lua-time-limit");
+        clusterClient.configSet(Collections.singletonMap("lua-time-limit", "100"), route).get();
 
         try (Script script = new Script(code, false);
                 GlideClusterClient testClient =
                         GlideClusterClient.createClient(
                                         commonClusterClientConfig()
-                                                .requestTimeout(15000)
+                                                .requestTimeout(10000)
                                                 .advancedConfiguration(
                                                         AdvancedGlideClusterClientConfiguration.builder()
                                                                 .connectionTimeout(10000)
@@ -3957,20 +3970,33 @@ public class CommandTests {
                                                 .build())
                                 .get()) {
 
-            // Poll scriptKill in background looking for "unkillable" error
-            CompletableFuture<Boolean> unkillableResult =
-                    pollForUnkillableInBackground(() -> clusterClient.scriptKill(route));
-
-            // Block on script execution to guarantee it's running on the server
+            // Start the script without blocking, so it runs on the server while we poll.
             CompletableFuture<Object> scriptFuture =
                     testClient.invokeScript(script, ScriptOptions.builder().key(key).build());
+
+            // Poll SCRIPT KILL until the server reports the script is unkillable (write script).
+            waitFor(
+                    () -> {
+                        try {
+                            clusterClient.scriptKill(route).get();
+                            return false; // returned OK - not the expected error, keep polling
+                        } catch (Exception e) {
+                            String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
+                            // NotBusy means the script hasn't started yet - keep polling.
+                            return msg.contains("unkillable");
+                        }
+                    },
+                    "Expected to find 'unkillable' error for write script");
+
             try {
                 scriptFuture.get();
             } catch (Exception ignored) {
                 // Write script completes normally after its duration
             }
-
-            assertTrue(unkillableResult.get(), "Expected to find 'unkillable' error for write script");
+        } finally {
+            clusterClient
+                    .configSet(Collections.singletonMap("lua-time-limit", originalLuaTimeLimit), route)
+                    .get();
         }
         waitForNotBusy(clusterClient::scriptKill);
         clusterClient.ping().get();
@@ -4012,55 +4038,6 @@ public class CommandTests {
                             result.completeExceptionally(
                                     new AssertionError(
                                             "Timed out waiting to kill script after " + SCRIPT_POLL_TIMEOUT_MS + "ms"));
-                        });
-        thread.setDaemon(true);
-        thread.start();
-        return result;
-    }
-
-    /**
-     * Polls scriptKill in a background thread until it returns an "unkillable" error, confirming the
-     * write script is running but cannot be killed.
-     */
-    private CompletableFuture<Boolean> pollForUnkillableInBackground(
-            Supplier<CompletableFuture<String>> killCommand) {
-        CompletableFuture<Boolean> result = new CompletableFuture<>();
-        Thread thread =
-                new Thread(
-                        () -> {
-                            long deadline = System.currentTimeMillis() + SCRIPT_POLL_TIMEOUT_MS;
-                            while (System.currentTimeMillis() < deadline) {
-                                try {
-                                    killCommand.get().get();
-                                } catch (Exception e) {
-                                    String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
-                                    if (msg.contains("unkillable")) {
-                                        result.complete(true);
-                                        return;
-                                    }
-                                    if (!msg.contains("no scripts in execution")
-                                            && !msg.contains("timed out")
-                                            && !msg.contains("timeout")) {
-                                        // Unexpected error - fail fast
-                                        result.completeExceptionally(e);
-                                        return;
-                                    }
-                                    // Expected: script hasn't started yet (NotBusy) or server is
-                                    // blocked by script (timeout) - keep polling
-                                }
-                                try {
-                                    Thread.sleep(SCRIPT_POLL_INTERVAL_MS);
-                                } catch (InterruptedException ie) {
-                                    result.completeExceptionally(ie);
-                                    Thread.currentThread().interrupt();
-                                    return;
-                                }
-                            }
-                            result.completeExceptionally(
-                                    new AssertionError(
-                                            "Timed out waiting for 'unkillable' error after "
-                                                    + SCRIPT_POLL_TIMEOUT_MS
-                                                    + "ms"));
                         });
         thread.setDaemon(true);
         thread.start();

